src/problem.jl

```diff
@@ -0,0 +1,104 @@
+abstract type AbstractSpecification end
+
+"""
+    struct Problem
+
+Program synthesis problem defined with a vector of [`AbstractSpecification`](@ref)s. 
+
+!!! warning
+    Please care that concrete `Problem` types with different values of `T` are never subtypes of each other. 
+"""
+struct Problem{T <: AbstractSpecification}
+    name::AbstractString
+    spec::T
+
+    function Problem(spec::T) where T<:AbstractSpecification
+        new{T}("", spec)
+    end
+end
+
+
+abstract type AbstractIOSpecification <: AbstractSpecification end
+
+"""
+    struct IOExample
+
+An input-output example.
+`in` is a [`Dict`](@ref) of `{Symbol,Any}` where the symbol represents a variable in a program.
+`out` can be anything.
+"""
+
+struct IOExample
+    in::Dict{Symbol, Any}
+    out::Any
+end
+
+"""
+    struct IOSpecification <: AbstractIOSpecification
+
+"""
+struct IOSpecification <: AbstractIOSpecification
+    examples::Vector{IOExample}
+end
+
+"""
+    struct IOMetricSpecification <: AbstractIOSpecification
+
+"""
+struct IOMetricSpecification <: AbstractIOSpecification
+    examples::AbstractVector{IOExample}
```

Comment:
this one is `AbstractVector` but `IOSpecification` uses `Vector`. We should be consistent